ambry-store/src/main/java/com.github.ambry.store/IndexSegment.java

```diff
@@ -543,6 +543,20 @@ void writeIndexSegmentToFile(Offset safeEndPoint) throws IOException, StoreExcep
           writer.writeShort(resetKey.getSecond().ordinal());
         }
 
+        // NOTE: In the event of a crash, it is possible that there is a part of the log that is not covered by the
+        // index. This happens due to the fact that a DELETE that occurs in the same segment as a PUT overwrites the
+        // PUT entry. Consider the following case:-
+        // (entries are of the form ID:TYPE:START_OFFSET-END_OFFSET)
+        // This is the order of operations
+        // A:PUT:0-100
+        // B:PUT:101-200
+        // A:DELETE:201-250
+        // These are the entries in the index segment
+        // A:DELETE:201-250
+        // B:PUT:101-200
+        // If safeEndPoint < 250, then B:PUT:101-200 will be written but not A:DELETE:201-250. If the process were to
+        // crash at this point, the index end offset would be 200 and the span 0-100 would not be represented in the
+        // index.
```

Comment:
I wanted to call out what could happen because of this - if this comment was around, I feel like the recovery bug would not have been introduced.